### ISSUE TITLE
https endpoints Linux Cluster

### ISSUE THREAD

**Issue by solidcloudio** [Azure/service-fabric-issues#611]
I've configured a secure Linux cluster. I'm able to connect to the Service Fabric Explorer with my Admin certificate.  Works fine under https.

I have several services I've configured, they all work fine using HTTP.

I've updated my configuration to use HTTPS

ServiceManifest
`<Resources>
    <Endpoints>
      
      <Endpoint Name="Label-AppTypeEndpoint" Port="8080" Protocol="https" Type="Input" />
    </Endpoints>
  </Resources>`

Application Manifest
`<ServiceManifestImport>
    <ServiceManifestRef ServiceManifestName="Label-AppPkg" ServiceManifestVersion="1.1.273" />
    <ConfigOverrides />
    <Policies>
      <ContainerHostPolicies CodePackageRef="Code">
        <RepositoryCredentials AccountName="obfuscated" Password="xxx" PasswordEncrypted="false"/>
        <PortBinding ContainerPort="8080" EndpointRef="Label-AppTypeEndpoint"/>
      </ContainerHostPolicies>
      <EndpointBindingPolicy CertificateRef="EndpointCert" EndpointRef="Label-AppTypeEndpoint"/>
    </Policies>
  </ServiceManifestImport>
<Certificates>
    <EndpointCertificate Name="EndpointCert" X509FindValue="83 46 a6 19 f0 ee 14 5e 59 c8 16 bb f1 e3 52 93 50 d3 6f 43" X509StoreName="MY"/>
  </Certificates>`

After deploying, my app still works under HTTP but HTTPS returns:  "This site can’t provide a secure connection"

My Admin and Server certificates are both Self-Signed certs, and I'm using the thumbprint for the Admin certificate in the X509FindValue.   This seems strange however, because my service is hosted under:

https://myclustername.westus2.cloudapp.azure.com:8070/health (not really a real url)

So it should be using the azure.com HTTPS certificate, not sure the correlation to the EndpointCertificate.

Documentation https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-service-manifest-resources#example-specifying-an-https-endpoint-for-your-service

"For Linux clusters, the MY store defaults to the folder /var/lib/sfcerts."

OK, I can see my Admin cert and Server cert there...  

Although nothing really says "What certificate" I'm supposed to use.  I'd assume once I get this on a custom domain that I'd use a cert I buy from a CA, but just using the azure defaults for now, what certificate does HTTPS use?

And why does HTTP still work while HTTPS doesnt when I specified in my configuration to use HTTPS?

**Comment by Mardoxx** [comment]
Weird how it's listening on http even though you have no htto endpoint specified.

>I'm using the thumbprint for the Admin certificate in the X509FindValue.
I believe there is a difference (perhaps I should learn this) between client and server certificates. By the sounds of it you are using your admin client certificate which won't have the correct intended purposes - so this may be why it's not working

**Comment by solidcloudio** [comment]
That's my point.  A "secure Linux cluster" uses an Admin Certificate (For accessing the dashboard) and a Server certificate ( so the nodes can talk to each other )  But neither of those certs should be the certificate the is used for HTTPS (SSL)

So thats the problem/question?  If you are using the Azure.com certificate, it wouldn't even use a "thumbprint"  It's not installed on my nodes..  So what does The Endpoint certificate entry in the Application Manifest have to do with using an SSL?  

In an Azure web-app, SSL is built in and available by the Azure infrastructure (unless I bought my own SSL) in which case I'd need to install it and I'd have a thumbprint for it.

**Comment by Mardoxx** [comment]
Someone else will fill you in better than I, but I believe there's some misunderstanding here. A [secure cluster](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-security) means that `node<->node` communication is secured -- it doesn't really have any bearing on what security, if any, the endpoints you expose use.

The instructions you linked [here](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-service-manifest-resources#example-specifying-an-https-endpoint-for-your-service) describe how to setup a service with an endpoint exposed which can terminate SSL traffic (or forward it). Somewhat unrelated to having a (in)secure cluster.

The certificate you provide to secure your endpoint should be different from the node server certificate (I presume this is so if this, more exposed, certificate is compromised, you don't end up compromising the security of your entire cluster). You're likely getting the "This site can’t provide a secure connection" because you're using a self signed certificate (or one which is invalid for use as a server certificate). Microsoft won't give you a certificate for `*.cloudapp.azure.com` so you'll have to either buy your own domain, get an SSL certificate for it, and point it to your cluster's IP address (or some other sort of configuration) .... or put up with the warnings!

I don't quite fully understand it myself, but hopefully I've given you enough information to assist.

Basically a secure cluster doesn't automagically make every endpoint you expose secure -- which I think is where the confusion lies?

**Comment by solidcloudio** [comment]
Yeah I get that.. I understand the cluster certificates and their uses:

Admin certificate - some cert to secure the Service Fabric Explorer access.
Server certificate - a cert the nodes to securely talk to each other.
(I've created both of these (self-signed) and they are working fine.

SSL certificate - normally a CA issued certificate that you install on a web server to secure inbound HTTPS traffic to the apps hosted on the nodes to the public.

I don't want to buy a certificate right now as this is just a test cluster, but my apps require HTTPS.  

If I had an app hosted at myclustername.westus2.cloudapp.azure.com:8070  for example, I can use
http://myclustername.westus2.cloudapp.azure.com:8070 by default, but I'd like to use https://myclustername.westus2.cloudapp.azure.com:8070

But this doesn't work.

So how do I do that latter?  I know I don't and wont have a certificate on my server for cloudapp.azure.com, Azure should provide that, just like a web app is accessible under http:// or https:// So how can I do this with a service fabric cluster?

The docs show using a CA issues cert, but I dont have one, I just want to use Azure's

**Comment by RajeetN** [comment]
The EndpointBinding in ApplicationManifest is only required on Windows so Service Fabric can configure HTTP.SYS with required certificate for users service. On Linux, Service Fabric does not need to do anything for https. Your app should listen on https://IPAddressOrFQDN:port. And you will need certificate with myclustername.westus2.cloudapp.azure.com common name, ServiceFabric/Azure do not provide that.

**Comment by solidcloudio** [comment]
So, if I have a certificate for myclustername.westus2.cloudapp.azure.com how is it configured?  Is it installed in the cluster as a secondary certificate?  You said: "On Linux, Service Fabric does not need to do anything for https. Your app should listen on https://IPAddressOrFQDN:port."  So, how do I configure this certificate if "Service Fabric doesnt need to do anything" ?

**Comment by Mardoxx** [comment]
So the config you had before is windows specific, SF configures HTTP.SYS (windows kernel level http listener) with provided cert (e.g. [here](https://stackoverflow.com/questions/4339861/cant-configure-http-sys-to-allow-ssl-trafic-netsh-exe)).

I [don't think](https://stackoverflow.com/questions/13421372/windows-http-sys-alternative-for-linux-platform) there is such an analogue in linux (no built in SSL), so instead common practice is you terminate your ssl connections at the application level --- most use wrappers around openSSL libraries I tihnk. So you'd have to look at whatever platform/framework you are using and see how to configure HTTPs there. E.g. [for kestrel in c# on .net core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?tabs=aspnetcore2x#how-to-use-kestrel-in-aspnet-core-apps) or [nodejs express.js](https://stackoverflow.com/questions/5998694/how-to-create-an-https-server-in-node-js). 

Hope that helps. I think that's correct-ish info!

**Comment by solidcloudio** [comment]
I guess this is similar to if we were running our apps in an IaaS setup with VMs?  Sure, in that context, I'd have to configure SSL at the VM, Apache, Tomcat level.  The only frustrating part about this is that now my apps will need an SSL embedded in the .WAR file, done at compilation time.  (Several of them to support multiple environments)

That's frustrating as this should be more of a DevOps excersize that I can implement in the hosting configuration rather than at the App level.  Seems to more tightly couple our apps to thier implementation environment.

Add to this that all of our apps are wrapped in environment agnostic docker containers.  I guess we'll have to configure the apps at a lower level.  This is frustrating as App services move this to a higher level and allow configuration of the hosting environment instead of further down the stack.

I really want to keep my apps "Environment Agnostic" - maybe I can do this somehow in the docker container configuration or the docker / app environment variables?

**Comment by RajeetN** [comment]
Maybe I am missing something here, why does app need cert embedded into it at build time? Have you looked at this? https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-securing-containers#certificate-management-for-containers

**Comment by solidcloudio** [comment]
I reviewed your link above.  Sorry, It really doesnt give me enough information.  

The article states:

For Linux clusters, the certificates(PEM) are simply copied over from the store specified by X509StoreName onto the container. The corresponding environment variables on linux are:

Certificates_ServicePackageName_CodePackageName_CertName_PEM
Certificates_ServicePackageName_CodePackageName_CertName_PrivateKey

The container service or process is responsible for importing the certificate files into the container. To import the certificate, you can use setupentrypoint.sh scripts or execute custom code within the container process. 

OK, so what is setupentrypoint.sh??  Seems this is a Jenkins script, which I'm not using Jenkins, so it really doesnt apply.  I tracked down this script under..  https://github.com/Azure-Samples/service-fabric-java-getting-started/blob/master/Services/JenkinsDocker/JenkinsSF/JenkinsOnSF/Code/setupentrypoint.sh

It doesn't appear to do anything with SSL certs...

So I understand the overall process.  IE I need to put a certificate on the SF Linux nodes and the somehow copy them into the container instance..   What runs this ?  When will this happen?  

I don't use Jenkins, I'm deploying Azure container images from PowerShell, so when do I ever have a chance to copy an SSL into the container ?

**Comment by solidcloudio** [comment]
@RajeetN Can you re-open tihs issue ?
